File: day1/day1.2.py
from typing import Generator


def rotate(start: int, shift: int, modulo: int) -> int:
    return (start + shift) % modulo


def num_passes(start: int, shift: int, modulo: int) -> int:
    res: int = abs(shift) // modulo
    end: int = rotate(start, shift, modulo)

    if start != 0 and (end == 0 or
       (start - end > 0 and shift > 0) or
       (start - end < 0 and shift < 0)):
        res += 1      
   
    return res


def get_rotations() -> Generator[int, None, None]:
    with open('input.txt', 'r') as input_file:
        for line in input_file:
            sign: int = 1 if 'R' == line[0] else -1

            yield sign * int(line[1:].strip())


def count_zeros(dial_size: int) -> int:
    total: int = 0
    position: int = 50

    for shift in get_rotations():
        total += num_passes(position, shift, dial_size)
        position = rotate(position, shift, dial_size)

    return total


if __name__ == '__main__':
    print(count_zeros(100))

File: day1/day1.2.cpp
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <print>
#include <string>
#include <vector>
#include "../common.h"

int STARTING_POSITION = 50;
int DEFAULT_DIAL_SIZE = 100;

[[nodiscard]] constexpr auto rotate(const int start, const int shift, const int modulo) -> int {
    const auto result = (start + shift) % modulo;
    return result < 0 ? result + modulo : result;
}

[[nodiscard]] constexpr auto num_passes(
    const int start,
    const int shift,
    const int modulo
) -> int {
    int res = std::abs(shift) / modulo;
    const int end = rotate(start, shift, modulo);

    if (start != 0 && (end == 0 ||
        (start - end > 0 && shift > 0) ||
        (start - end < 0 && shift < 0)))
        res += 1;

    return res;
}

[[nodiscard]] auto parse_rotation(const std::string_view line) -> int {
    const int sign = line.front() == 'R' ? 1 : -1;
    return sign * std::stoi(std::string{line.substr(1)});
}

[[nodiscard]] auto get_rotations() -> std::vector<int> {
    std::vector<int> rotations;

    {
        auto f = AOC_INPUT();
        std::string line;
        while (std::getline(f, line))
            rotations.push_back(parse_rotation(line));
    }

    return rotations;
}

[[nodiscard]] auto count_zeros(const int dial_size) -> int {
    int total = 0;
    int position = STARTING_POSITION;

    for (const int shift : get_rotations()) {
        total += num_passes(position, shift, dial_size);
        position = rotate(position, shift, dial_size);
    }

    return total;
}

auto main() -> int {
    std::println("{}", count_zeros(DEFAULT_DIAL_SIZE));

    return 0;
}
